pkg/ebpf/c/tracee.bpf.c

```diff
@@ -585,21 +585,55 @@ int tracepoint__sched__sched_process_fork(struct bpf_raw_tracepoint_args *ctx)
 // number of iterations - value that the verifier was seen to cope with - the higher, the better
 #define MAX_NUM_MODULES 600
 
+enum
+{
+    PROC_MODULES = 1 << 0,
+    KSET = 1 << 1,
+    MOD_TREE = 1 << 2,
+    NEW_MOD = 1 << 3,
+    FULL_SCAN = 1 << 30,
+    HIDDEN_MODULE = 1 << 31,
+};
+
 // This map is only used from kernels 5.2 and above.
 // If we won't create the map, the userspace will receive an error when interacting with it,
 // and since the userspace is not kernel version aware, it won't know whether it's because
 // the map simply wasn't created due to kernel version, or something unexpected failed.
 BPF_HASH(modules_map, u64, kernel_module_t, MAX_NUM_MODULES);
-
-void __always_inline send_hidden_module(u64 mod_addr, char *mod_name, program_data_t *p)
+BPF_HASH(new_module_map, u64, kernel_new_mod_t, MAX_NUM_MODULES);
+
+// We only care for modules that got deleted or inserted between our scan and if we detected
+// something suspicious. Since it's a very small time frame, it's not likely that a large amount of
+// modules will be deleted. Instead of saving a map of deleted modules, we could have saved the last
+// deleted module timestamp and if we detected something suspicious, verify that no modules got
+// deleted between our check. This is preferable space-wise (u64 instead of a map), but an attacker
+// might start unloading modules in the background and race with the check in order to abort
+// reporting for hidden modules.
+BPF_LRU_HASH(recent_deleted_module_map, u64, kernel_deleted_mod_t, 50);
+BPF_LRU_HASH(recent_inserted_module_map,
+             u64,
+             kernel_new_mod_t,
+             50); // Likewise for module insertion
+
+void __always_inline lkm_seeker_send_to_userspace(struct module *mod, u32 *flags, program_data_t *p)
 {
     reset_event_args(p);
+    u64 mod_addr = (u64) mod;
+    char *mod_name = mod->name;
+    const char *mod_srcversion = READ_KERN(mod->srcversion);
```

Comment:
READ_KERN was dropped, should use BPF_CORE_READ (and similar) macros (directly from libbpf UAPI headers).